### PR TITLE
Kpi resources

### PR DESCRIPTION
**Pull request by azizullahsaeidi**
1. Get total resouces base on language -> filter by `language`, `date` and `gender`
2. Get top 10 downloaded resources -> filter by `language` and `date`
3. Get top 10 favorite resources  -> filter by `language` and `date`
4. Get top 10 downloaded resources by file size  -> filter by `language`, `date` and `gender`
5. Get total downloaded File Sizes
6. Get top 10 authors -> filter by `language` and `date`
7. Get top 10 publishers -> filter by `language` and `date`

Please run below command
```
php artisan migrate --path=/database/migrations/2024_04_09_045125_add_index_to_download_counts_table.php
```

### CODE REVIEW

**Comment by uyscuti-wiki** [comment]
@azizullahsaeidi This looks great overall. Few comments about the design:

- Let's replace these tags with circles with three (or two or one depending on what's supported) colour fills next to the title in the card header image 
Next to the filter button, aligned right perhaps, you can signify what each coloured circle means. For instance
● - Date
● - Gender
● - Language
Then, only display the circles that are supported in every card header. 
- Where possible, link resources and filters. 
- Where no data is returned, display '[not available]'.
- Replace 'Resouces base on Language' with 'Total resources by languages'.
- Modify 'Top 10 favorite resources' to 'Top 10 favorited resources'.
- Replace 'Downloaded File Sizes' with 'Total data downloaded'.
- Modify 'Top 10 Authors' to 'Top 10 authors'.
- Modify 'Top 10 Publisher' to 'Top 10 publishers'.

I'll add more as new comments as I continue to review the code. Thanks!

File: app/Traits/DownloadCountTrait.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+namespace App\Traits;
+
+use App\Models\DownloadCount;
+
+trait DownloadCountTrait
```
Comment (uyscuti-wiki) [inline app/Traits/DownloadCountTrait.php:7]: Has this trait been used anywhere?

File: app/Traits/GenderTrait.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+use Mcamara\LaravelLocalization\Facades\LaravelLocalization;
+
+trait GenderTrait
+{
+    public function genders(): array
+    {
+        return [
+            'Male' => 'Male',
+            'Female' => 'Female',
+            'None' => 'None'
```
Comment (uyscuti-wiki) [inline app/Traits/GenderTrait.php:15]: Could we stick to lowercase initial character for the keys?

**Review by uyscuti-wiki — CHANGES_REQUESTED** [review]
Thank you for making the changes. Looks pretty good to me. I have requested some minor changes below. Could we also add unit tests?

File: app/Traits/GenderTrait.php (hunk 2 above)
Comment (azizullahsaeidi) [reply app/Traits/GenderTrait.php]: Yes sure. Thank you

File: app/Http/Controllers/ResourceAnalyticsController.php [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+        $authors = $this->getTop10AuthorsOrPublishers($request, 'resource_authors'); // Get top 10 authors
+        $totalResources = $this->getTotalResouceBaseOnLanguage($request); // Total Resources base on Language
+        $top10FavoriteResources = $this->getTop10FavoriteResources($request); // Get top 10 favorite resources
+        $publishers = $this->getTop10AuthorsOrPublishers($request, 'resource_publishers'); // Get top 10 publishers
+        $top10DownloadedResources = $this->getTop10DownloadedResources($request); // Get top 10 downloaded resources
+        $top10DownloadedResourcesByFileSizes = $this->getTop10DownloadedResourcesByFileSize($request); // Get top 10 downloaded resources by file size
+        $sumOfAllIndividualDownloadedFileSizes = $this->getSumOfAllIndividualDownloadedFileSizes(); // Sum of all individual downloaded file sizes
+
+        return view('admin.analytics.resource-analytics.index', compact(['records', 'genders', 'languages', 'reportType', 'totalResources', 'sumOfAllIndividualDownloadedFileSizes', 'authors', 'publishers', 'top10DownloadedResources', 'top10DownloadedResourcesByFileSizes', 'top10FavoriteResources']));
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Http/Controllers/ResourceAnalyticsController.php]: Removed. They not defined. Thank you Aaron

File: app/Http/Controllers/ResourceAnalyticsController.php [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+    public function index(Request $request): View
+    {
+        $genders = $this->genders();
+        $languages = $this->getLanguages();
+
+        $authors = $this->getTop10AuthorsOrPublishers($request, 'resource_authors'); // Get top 10 authors
+        $totalResources = $this->getTotalResouceBaseOnLanguage($request); // Total Resources base on Language
+        $top10FavoriteResources = $this->getTop10FavoriteResources($request); // Get top 10 favorite resources
+        $publishers = $this->getTop10AuthorsOrPublishers($request, 'resource_publishers'); // Get top 10 publishers
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Http/Controllers/ResourceAnalyticsController.php]: Done

File: app/Http/Controllers/ResourceAnalyticsController.php [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+class ResourceAnalyticsController extends Controller
+{
+    use LanguageTrait, GenderTrait;
+
+    public function index(Request $request): View
+    {
+        $genders = $this->genders();
+        $languages = $this->getLanguages();
+
+        $authors = $this->getTop10AuthorsOrPublishers($request, 'resource_authors'); // Get top 10 authors
```
[the remark this answers is not in the document]
Comment (azizullahsaeidi) [reply app/Http/Controllers/ResourceAnalyticsController.php]: Done

**Comment by azizullahsaeidi** [comment]
> Thank you for making the changes. Looks pretty good to me. I have requested some minor changes below. Could we also add unit tests?

@uyscuti-wiki if you agree, let me write the test in a separate PR.

File: app/Traits/DownloadCountTrait.php (hunk 1 above)
Comment (uyscuti-wiki) [reply app/Traits/DownloadCountTrait.php]: @azizullahsaeidi could you clarify?

Comment (azizullahsaeidi) [reply app/Traits/DownloadCountTrait.php]: @uyscuti-wiki  I already deleted this trait.

**Review by uyscuti-wiki — APPROVED** [review]
Looks great! Thank you for your efforts, Aziz.

**Pull request merged**